src/vm/win32threadpool.cpp

```diff
@@ -110,7 +110,7 @@ int ThreadpoolMgr::ThreadAdjustmentInterval;
 #define INVALID_HANDLE ((HANDLE) -1)
 #define NEW_THREAD_THRESHOLD            7       // Number of requests outstanding before we start a new thread
 #define CP_THREAD_PENDINGIO_WAIT 5000           // polling interval when thread is retired but has a pending io
-#define GATE_THREAD_DELAY 500 /*milliseconds*/
+#define GATE_THREAD_DELAY 200 /*milliseconds*/
```

Comment:
The tweaks to timing of pool expansion are mostly to deal with the `SslStream_SameCertUsedForClientAndServer_Ok` testcase. There are certainly better ways. 
Consider this not be a part of the change.

---
In reply to: [250333993](https://github.com/dotnet/coreclr/pull/18403#discussion_r250333993) [](ancestors = 250333993)

Reply:
The test is in FX repo, so it cannot be in the same PR. 
Also I am not sure what would be the "right" fix. Ideally we should be tolerant to that kind of starvation and not ask users to change code. It would feel a bit of giving up if I fix the test :-) 
May not have a lot of choice in the short term though.

---
In reply to: [250437036](https://github.com/dotnet/coreclr/pull/18403#discussion_r250437036) [](ancestors = 250437036)

Reply:
Currently, the thread pool is not designed to handle long-blocking work or work that creates newer dependencies in the same thread pool. Both of these scenarios have fundamental issues that even a perfect thread pool cannot solve, but they can be handled much better than they are now (would involve being able to detect blocking and have the opportunity to respond immediately to blocking). For now I think a reasonable fix would be to eliminate any dependency on the thread pool for the test and to instead use separate threads, or if tasks are being used they can also be flagged as long-running.

Reply:
I'm not sure yet where the starvation is coming from in the test, I don't expect to see any sort of starvation from the test code, so it is a bit odd, I take it back.

Reply:
It is not a single test. The whole area uses the following for network simulation - [https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Common/tests/System/Net/VirtualNetwork/VirtualNetwork.cs](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Common/tests/System/Net/VirtualNetwork/VirtualNetwork.cs)
Some tests intentionally test timeouts where it is _expected_ that server or client gets no response. These tests run asyncronously, so when enough of them run, there are definitely blocked workers if you check in the debugger. 

`SslStream_SameCertUsedForClientAndServer_Ok` is not special in any ways other than it has shorter timeout than the rest of similar tests - 15sec vs. 30sec. so it fails more often. It failed occasionally with regular threadpool too, but recently it mostly passes. See [https://github.com/dotnet/corefx/issues/23291](https://github.com/dotnet/corefx/issues/23291)

Why this does not repro locally? - My guess is that CI machines are lower spec VMs and starvation issues have more impact.

NOTE: I do not have 100% confirmation that this is the reason for failures, it is possible that there is another issue, but I think there are good indications that this is the case.

---
In reply to: [250635147](https://github.com/dotnet/coreclr/pull/18403#discussion_r250635147) [](ancestors = 250635147)

Reply:
Ah ok, maybe the blocking tasks used for async operations in `VirtualNetworkStream` could be flagged as long-running